runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingStepMetricsContainer.java

```diff
@@ -68,6 +74,14 @@ public class StreamingStepMetricsContainer implements MetricsContainer {
   private MetricsMap<KV<MetricName, HistogramData.BucketType>, HistogramCell> perWorkerHistograms =
       new MetricsMap<>(HistogramCell::new);
 
+  private Map<MetricName, Instant> perWorkerCountersByFirstStaleTime = new ConcurrentHashMap<>();
+
+  // PerWorkerCounters that have been longer than this value will be removed from the underlying
+  // metrics map.
+  private Duration maximumPerWorkerCounterStaleness = Duration.ofMinutes(5);
+
+  private Clock clock = Clock.systemUTC();
```

Comment:
see testing comment below, if you inject these then they can call be final since they won't need to be reassigned.